sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala

```diff
@@ -294,7 +296,12 @@ object ShuffleExchangeExec {
           sorter.sort(iter.asInstanceOf[Iterator[UnsafeRow]])
         }
       } else {
-        rdd
+        part match {
+          case partitioner: RangePartitioner[InternalRow @unchecked, _]
```

Comment:
Do we need to match type argument of `RangePartitioner`?

Maybe just:

```scala
part match {
  case partitioner: RangePartitioner[_, _] if partitioner.getSampledArray != null =>
    ...
  case _ => rdd
}
```

Reply:
```
def prepareShuffleDependency(
      rdd: RDD[InternalRow],
      outputAttributes: Seq[Attribute],
      newPartitioning: Partitioning,
      serializer: Serializer,
      sparkContext: SparkContext = null): ShuffleDependency[Int, InternalRow, InternalRow] = {
```
the rdd is defined as RDD[InternalRow], so we must match type argument here

Reply:
I think we don't get `RangePartitioner` with other type argument than `InternalRow`, right?

Reply:
yes, but 
```
def parallelize[T: ClassTag](
      seq: Seq[T],
      numSlices: Int = defaultParallelism): RDD[T] = withScope {
    assertNotStopped()
    new ParallelCollectionRDD[T](this, seq, numSlices, Map[Int, Seq[String]]())
  }
```
the parallelize function needs this ClassTag, so we must match it here.
I tried to match RangePartitioner[_, _], but there is an error:
```
Error:(302, 37) No ClassTag available for _
Error occurred in an application involving default arguments.
            sparkContext.parallelize(partitioner.getSampledArray.toSeq, rdd.getNumPartitions)
```

Reply:
Ok. I see.